### PR TITLE
feat: added generic proto file for grpc service

### PR DESCRIPTION
**Pull request by Proto007**
## What does this PR address?

The generic proto file contains a `BentoML` service with an RPC called `call`. Fixed `io_descriptor.proto` file with updated name of `Array` instead of `NumpyNdarray`.

## Before submitting:

- [X] Does the Pull Request follow [Conventional Commits specification](https://www.conventionalcommits.org/en/v1.0.0/#summary) naming? Here are [GitHub's
guide](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request) on how to create a pull request.
- [X] Does the code follow BentoML's code style, both `make format` and `make lint` script have passed ([instructions](https://github.com/bentoml/BentoML/blob/main/DEVELOPMENT.md#style-check-auto-formatting-type-checking))?
- [X] Did you read through [contribution guidelines](https://github.com/bentoml/BentoML/blob/main/CONTRIBUTING.md#ways-to-contribute) and follow [development guidelines](https://github.com/bentoml/BentoML/blob/main/DEVELOPMENT.md#start-developing)?
- [ ] Did your changes require updates to the documentation? Have you updated
  those accordingly? Here are [documentation guidelines](https://github.com/bentoml/BentoML/tree/main/docs) and [tips on writting docs](https://github.com/bentoml/BentoML/tree/main/docs#writing-documentation).
- [ ] Did you write tests to cover your changes?

## Who can help review?
The proto file and changes are based on our discussion. Would appreciate any feedback! 
@aarnphm @sauyon @ssheng

### CODE REVIEW

File: protos/bentoml_service.proto [hunk 1]
```diff
@@ -0,0 +1,29 @@
+syntax = "proto3";
```
Comment (aarnphm) [inline protos/bentoml_service.proto:1]: This file should just be named `service.proto`, as this is bentoml spec only. Service is a core bentoml concept, hence no need for the distinction here.

File: protos/io_descriptors.proto [hunk 2]
```diff
@@ -1,52 +1,51 @@
...
+	repeated uint64 uint64_ = 5;
+	repeated sint32 sint32_ = 6;
+	repeated sint64 sint64_ = 7;
+	repeated google.protobuf.Timestamp timestamp_ = 8;
+	repeated google.protobuf.Duration duration_ = 9;
+	repeated bool bool_ = 10;
+	repeated string string_ = 11;
+	repeated bytes bytes_ = 12;
+	repeated Array array_ = 13;
+	repeated Tuple tuple_ = 14;
```
Comment (aarnphm) [inline protos/io_descriptors.proto:50]: How about implementing an interable type. This encapsulate both tuple, list, and array @ssheng

File: protos/bentoml_service.proto (hunk 1 above)
Comment (Proto007) [reply protos/bentoml_service.proto]: Done!

**Review by aarnphm — COMMENTED** [review]
A few early things:
- Make sure to follow convention: We should adhere to [Google convention](https://developers.google.com/protocol-buffers/docs/style#enums).
- Comments should be consistent. Preferrably `//`

File: protos/service.proto [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+}
+
+/* BentoServiceInput represents the output for a grpc response. Uses `BentoServiceMessage` to represent response data. */
+// TODO status, error
+message BentoServiceOutput {
+	BentoServiceMessage output = 1;
+}
+
+/* BentoML service with rpc `call()` that takes a `BentoServiceInput` and returns `BentoServiceOutput` */
+service BentoML{
```
[the remark this answers is not in the document]
Comment (Proto007) [reply protos/service.proto:27]: Based on https://developers.google.com/protocol-buffers/docs/style#services , changing RPC from `call()` to `Call()`.

[the remark this answers is not in the document]
Comment (Proto007) [reply protos/service.proto]: Done

[the remark this answers is not in the document]
Comment (aarnphm) [reply protos/service.proto]: Change it to `RouteCall`

File: protos/service.proto [hunk 4]
```diff
@@ -0,0 +1,29 @@
+syntax = "proto3";
+
+import "io_descriptors.proto";
+
+/* Payload allows single instance of one `io_descriptor`. */
+message Payload {
+	oneof io_descriptor{
```
Comment (aarnphm) [inline protos/service.proto:7]: can you also add a script to proto formatter?

File: protos/service.proto (hunk 3 above)
[the remark this answers is not in the document]
Comment (Proto007) [reply protos/service.proto]: On it.

File: protos/io_descriptors.proto [hunk 5]
```diff
@@ -1,52 +1,51 @@
 syntax = "proto3";
 
-import "google/protobuf/timestamp.proto";
 import "google/protobuf/duration.proto";
+import "google/protobuf/timestamp.proto";
 
-/* Value represents a single instance of the supported datatypes. */
-// TODO: int32, int64, fixed32, fixed64, sfixed32, sfixed64
+/* Value represents a single instance of the supported datatypes. 
+ * TODO: int32, int64, fixed32, fixed64, sfixed32, sfixed64 */
```
[the remark this answers is not in the document]
Comment (Proto007) [reply protos/io_descriptors.proto:7]: I was following https://developers.google.com/protocol-buffers/docs/proto#adding_comments. Should I make the change?

File: protos/service.proto [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+/* `BentoServiceResponse` represents a grpc response. Uses `Payload` to represent response data. 
+ * TODO: status, error */
+message BentoServiceResponse {
+	Payload output = 1;
+}
+
+/* `BentoService` with rpc `Call()` that takes a `BentoServiceRequest` and returns `BentoServiceResponse` */
+service BentoService {
+	rpc Call(BentoServiceRequest) returns (BentoServiceResponse);
```
Comment (aarnphm) [inline protos/service.proto:28]:
```suggestion
	rpc RouteCall(BentoServiceRequest) returns (BentoServiceResponse);
```

File: protos/service.proto [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+	Payload input = 2;
+}
+
+/* `BentoServiceResponse` represents a grpc response. Uses `Payload` to represent response data. 
+ * TODO: status, error */
+message BentoServiceResponse {
+	Payload output = 1;
+}
+
+/* `BentoService` with rpc `Call()` that takes a `BentoServiceRequest` and returns `BentoServiceResponse` */
```
Comment (aarnphm) [inline protos/service.proto:26]:
```suggestion
// `BentoService` with rpc `RouteCall()` that takes a `BentoServiceRequest` and returns `BentoServiceResponse`
```

File: protos/service.proto [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+
+/* `BentoServiceRequest` represents a grpc request. Uses `Payload` to represent request data.
+ * @api_name: name of the api user wants to call using the provided `Payload` */
+message BentoServiceRequest {
+	string api_name = 1;
+	Payload input = 2;
+}
+
+/* `BentoServiceResponse` represents a grpc response. Uses `Payload` to represent response data. 
+ * TODO: status, error */
```
Comment (aarnphm) [inline protos/service.proto:21]:
```suggestion
// `BentoServiceResponse` represents a grpc response. Uses `Payload` to represent response data. 
// TODO: status, error
```

File: protos/service.proto [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+/* Payload allows single instance of one `io_descriptor`. */
+message Payload {
+	oneof io_descriptor{
+		string text = 1;
+		Array array = 2;
+	}
+}
+
+/* `BentoServiceRequest` represents a grpc request. Uses `Payload` to represent request data.
+ * @api_name: name of the api user wants to call using the provided `Payload` */
```
Comment (aarnphm) [inline protos/service.proto:14]:
```suggestion
// `BentoServiceRequest` represents a grpc request. Uses `Payload` to represent request data.
// @api_name: name of the api user wants to call using the provided `Payload`
```

File: protos/io_descriptors.proto [hunk 10]
```diff
@@ -1,52 +1,51 @@
...
-	repeated google.protobuf.Timestamp timestamp_=8;
-	repeated google.protobuf.Duration duration_=9;
-	repeated bool bool_=10;
-	repeated string string_=11;
-	repeated bytes bytes_=12;
-	repeated NumpyNdarray array_=13;
-	repeated Tuple tuple_=14;
+/* Array contains a dtype which identifies the type of the array.
+ * The repeated field for the identified dtype contains the array. 
+ * TODO: int32, int64, fixed32, fixed64, sfixed32, sfixed64 */
```
Comment (aarnphm) [inline protos/io_descriptors.proto:35]:
```suggestion
// Array contains a dtype which identifies the type of the array.
// Repeated field for the identified dtype contains the array. 
// TODO: int32, int64, fixed32, fixed64, sfixed32, sfixed64
```

File: protos/io_descriptors.proto [hunk 11]
```diff
@@ -1,52 +1,51 @@
...
 
-/* Tuple represents a repeated field containing data with same or different datatypes */
+/* Tuple represents a repeated field containing data with same or different datatypes. */
 message Tuple{
-	repeated Value value_=1;
+	repeated Value value_ = 1;
 }
 
-// TODO: message complex types
+/* TODO: message complex types */
```
Comment (aarnphm) [inline protos/io_descriptors.proto:31]:
```suggestion
// TODO: message complex types
```

File: protos/io_descriptors.proto [hunk 12]
```diff
@@ -1,52 +1,51 @@
...
+		bool bool_ = 9;
+		string string_ = 10;
+		bytes bytes_ = 11;
+		Array array_ = 13;
+		Tuple tuple_ = 14;
 	}
 }
 
-/* Tuple represents a repeated field containing data with same or different datatypes */
+/* Tuple represents a repeated field containing data with same or different datatypes. */
```
Comment (aarnphm) [inline protos/io_descriptors.proto:26]:
```suggestion
// Tuple represents a repeated field containing data with same or different datatypes.
```

File: protos/io_descriptors.proto [hunk 13]
```diff
@@ -1,52 +1,51 @@
...
-import "google/protobuf/timestamp.proto";
 import "google/protobuf/duration.proto";
+import "google/protobuf/timestamp.proto";
 
-/* Value represents a single instance of the supported datatypes. */
-// TODO: int32, int64, fixed32, fixed64, sfixed32, sfixed64
+/* Value represents a single instance of the supported datatypes. 
+ * TODO: int32, int64, fixed32, fixed64, sfixed32, sfixed64 */
 message Value{
 	oneof dtype{
```
Comment (aarnphm) [inline protos/io_descriptors.proto:9]:
```suggestion
	oneof dtype {
```

File: protos/io_descriptors.proto (hunk 5 above)
Comment (aarnphm) [inline protos/io_descriptors.proto:7]:
```suggestion
// Value represents a single instance of the supported datatypes. 
// TODO: int32, int64, fixed32, fixed64, sfixed32, sfixed64.
```

**Review by aarnphm — APPROVED** [review]

File: protos/service.proto (hunk 3 above)
[the remark this answers is not in the document]
Comment (sauyon) [reply protos/service.proto]: Why `RouteCall`?

File: protos/service.proto [hunk 14]
```diff
@@ -0,0 +1,29 @@
+syntax = "proto3";
+
+import "io_descriptors.proto";
+
+/* BentoServiceMessage allows single instance of one `io_descriptor`. */
+message BentoServiceMessage {
+	oneof io_descriptor{
+		string text = 1;
+		Array array = 2;
+	}
```
[the remark this answers is not in the document]
Comment (sauyon) [reply protos/service.proto:10]: We should reserve a range eventually, yes.

File: protos/io_descriptors.proto (hunk 5 above)
[the remark this answers is not in the document]
Comment (sauyon) [reply protos/io_descriptors.proto:7]: I don't really see the problem with block comments?

File: protos/io_descriptors.proto (hunk 2 above)
Comment (aarnphm) [reply protos/io_descriptors.proto:50]: fair point.

**Pull request merged**